### DISCUSSION TITLE
Is inventory seeder (drop from inventory) an easy hack to make?

### DISCUSSION THREAD

**Discussion by SirEnder125** [RedShyGuy/Vapecord-ACNL-Plugin#26] [Q&A] [answered]
Hello. I recently made a text to item hack, yet it does not let you spawn trees, rocks, etc.
How can I do this? Is it difficult? Do I need to know ASM? Do I need a debugger?

Any help is appreciated. Thanks.

EDIT: Thanks for all the help! :wink:

**Comment by RedShyGuy** [comment]
The game has a check which checks the inventory if it has invalid items in it, if there is the icon of it will not be showed (the item itself will still be there). You would need to disable that check through assembly. Though to make it easier for you I can send you my way of doing it in 30 minutes!

**Reply by SirEnder125** [reply]
Really? Thanks! :P

**Comment by RedShyGuy** [comment, marked as answer]
This will enable the items from showing in your inventory
```cpp
bool InvalidItemStop(u32* Item) {
//you could add a check here which checks what items can be valid and what not, just use the "Item" argument 
	return 1; //makes all items valid
}

void YourEntry(MenuEntry *entry) {
	static Hook YourHook;
	u32 AddressToHookOn = 0x72511C;
	if(entry->WasJustActivated()) {
		YourHook.Initialize(AddressToHookOn, (u32)InvalidItemStop);
		YourHook.SetFlags(USE_LR_TO_RETURN); //might need to change that, as I use a different version of CTRPF
		YourHook.Enable();
	}
}
```
But due to the seed items (trees, bushes) not getting any settings (settings are like Drop, Show Off, etc) assigned you will not be able to drop them, this patch here will add the ability to drop them:
```cpp
	Process::Write32(0x19C050, 0xE1A00000);
```
If your plugin does not use ACNL USA Orig 1.5 I can also give you the address for any other region if you need it!
This should be everything you need!

**Reply by SirEnder125** [reply]
Well, I am a bit confused.
1. how does InvalidItemStop(u32* item) do anything if it just returns 1?
2. What is a hook and why do I need to use it?

Sorry if these are noob questions, I am quite the noob. :/

**Reply by RedShyGuy** [reply]
Soo
1. What are hooks
Hooks are used to "hook" at code from the game with a custom function, so if the in-game function is called by the game your custom function will be called instead!
2. Why does InvalidItemStop do anything
Well InvalidItemStop is a custom function that is hooked at a specific position, that position calls a in-game function which will check if an item is valid, if it is valid it will return 1 (true) and if not it will return 0 (false). So this custom function just makes it always return 1, which means every item is valid!

**Reply by SirEnder125** [reply]
Oh, ok. :) Well, the "YourEntry" function, where do I put it? At the beggining of one of my cheats? Or what?

**Reply by RedShyGuy** [reply]
You do not need to use that by itself, you can just copy out the content of that entry to one of yours

**Reply by SirEnder125** [reply]
Oh, so I just paste that into my Text to Item, for instance?

**Reply by SirEnder125** [reply]
Just out of curiosity, where do I put this line?
> Process::Write32(0x19C050, 0xE1A00000);

Also, the SetFlags function is invalid, so is the USE_LR_TO_RETURN... :/

Got to go for a bit.

**Reply by RedShyGuy** [reply]
> Just out of curiosity, where do I put this line?

Wherever you want

> Also, the SetFlags function is invalid, so is the USE_LR_TO_RETURN... :/

Yea I figured that, this should work then:
```cpp
YourHook.flags.ExecuteOverwrittenInstructionBeforeCallback  = false;
````
replace the SetFlags line with this ^

**Reply by SirEnder125** [reply]
Alright. Thanks for the info. It helped! :D

**Reply by SirEnder125** [reply]
I can't compile yet due to this error: "An exception occurred: yaml-cpp: error at line 10, column 23: illegal map value"

**Reply by RedShyGuy** [reply]
Huh, what line is it?

**Reply by SirEnder125** [reply]
I'm not sure, I don't even know what file it is talking about.

**Reply by RedShyGuy** [reply]
Hm, I think I added one space too much in the new thing I sent you, remove the space (so it looks like this:)
```cpp
YourHook.flags.ExecuteOverwrittenInstructionBeforeCallback = false;
```
Though I am not sure if that would even count as an error

**Reply by RedShyGuy** [reply]
Ah wait you seem to have yaml-cpp (it is this it seems like: https://cppsecrets.com/users/5577106971051101049711411510410511653495664103109971051084699111109/YAML-CPP-Introduction.php). Really not sure what that is and why you have that but it caused the issue.

**Reply by SirEnder125** [reply]
Huh, I don't know what that is either. Where
would it be located?

**Reply by RedShyGuy** [reply]
Should be inside your source I believe, I really don't know anything about it. But it is weird that you didn't get that error earlier, are you sure you didn't change anything else?

**Reply by SirEnder125** [reply]
No, I had just added the invalid item stop function and tried to compile.
I don't recall doing anything else to my source. :/

**Reply by RedShyGuy** [reply]
Try removing it and seeing if the error is still there (would be super weird if it wouldnt)

**Reply by SirEnder125** [reply]
Removing the function? I have already commented it out and tried to build, same error.

**Reply by RedShyGuy** [reply]
Hm okay good, can you somehow sent me your Source Code? (i dont know if something like private sharing even exists on github lol)
I could see if the error comes directly from the Source

**Reply by SirEnder125** [reply]
Ok, I downloaded CTRPF and moved my source to the downloaded folder and it now compiles.
I guess my former CTRPF download was messed up (probably by myself. lol)

**Reply by RedShyGuy** [reply]
Ah thats great 👍

**Reply by SirEnder125** [reply]
Indeed. :P Now I will find my 2ds and test the cheat.

**Reply by SirEnder125** [reply]
Ok, I think something went wrong, I tried to spawn a Christmas tree and it just spawned nothing.

**Reply by SirEnder125** [reply]
Here is my text to item code:
```C++
bool InvalidItemStop(u32* Item) 
{
	// You could add a check here which checks what items can be valid and what not, just use the "Item" argument.
	return 1; // Makes all items valid.
}

void Text2Item(MenuEntry* entry)
{
	if (entry->Hotkeys[0].IsDown())
	{
		static Hook YourHook;
		u32 AddressToHookOn = 0x72511C;
		if (entry->WasJustActivated()) {
			YourHook.Initialize(AddressToHookOn, (u32)InvalidItemStop);
			YourHook.flags.ExecuteOverwrittenInstructionBeforeCallback = false;
			YourHook.Enable();
		}
		Process::Write32(0x19C050, 0xE1A00000);

		Keyboard keyboard("What item would you like?");
		keyboard.SetMaxLength(8);
		u32 itemid;

		if (keyboard.Open(itemid) != -1)
		{
			Process::Write32(INVENTORY_ADDR, itemid);
		}
	}
}
```

**Reply by RedShyGuy** [reply]
Do you use USA 1.5 ? (If you dont know send me your inventory address and then I will know)

**Reply by SirEnder125** [reply]
I'm pretty sure that I am using that. And it does change the value, it just can't be a tree/rock. etc.
Inv Address: 0x31F2DBF0

**Reply by RedShyGuy** [reply]
okay yea thats the correct one, hmm and regular items work?

**Reply by SirEnder125** [reply]
Yes, I just spawned 99,000 bells. It worked.

**Reply by RedShyGuy** [reply]
Oh oops I forgot something haha sorry.
Add this and it will work:
```cpp
Process::Write32(0x76A894, 0xE3A00001);
```

**Reply by RedShyGuy** [reply]
You wouldn't even need the hook part with this, I think I used it for something else not really needed for this sorry lol.
So to be exact you only need:
```cpp
Process::Write32(0x19C050, 0xE1A00000);
Process::Write32(0x76A894, 0xE3A00001);
```

**Reply by SirEnder125** [reply]
Ok, so the code should be:
```C++
bool InvalidItemStop(u32* Item) 
{
	// You could add a check here which checks what items can be valid and what not, just use the "Item" argument.
	return 1; // Makes all items valid.
}

void Text2Item(MenuEntry* entry)
{
	if (entry->Hotkeys[0].IsDown())
	{
		Process::Write32(0x19C050, 0xE1A00000);
		Process::Write32(0x76A894, 0xE3A00001);

		Keyboard keyboard("What item would you like?");
		keyboard.SetMaxLength(8);
		u32 itemid;

		if (keyboard.Open(itemid) != -1)
		{
			Process::Write32(INVENTORY_ADDR, itemid);
		}
	}
}
```
Correct?

**Reply by RedShyGuy** [reply]
To make it a bit more better I would do:
```cpp
void Text2Item(MenuEntry* entry)
{
	if(entry->WasJustActivated()) 
	{
		Process::Write32(0x19C050, 0xE1A00000);
		Process::Write32(0x76A894, 0xE3A00001);
	}

	if (entry->Hotkeys[0].IsDown())
	{
		Keyboard keyboard("What item would you like?");
		//keyboard.SetMaxLength(8); not needed, setmaxlenght only works for string keyboards
		u32 itemid;

		if (keyboard.Open(itemid) != -1)
		{
			Process::Write32(INVENTORY_ADDR, itemid);
		}
	}
}
```
This way it will only write the patch to the code once when you enable the entry instead of always when you press the buttons!
(I also just figured out that the hook way also works, you just misplaced it and I didn't notice it lol, but that doesn't matter as this will work just fine)

**Reply by SirEnder125** [reply]
Oh, ok. Thanks. Do I also need to keep the InvalidItemStop() function?

**Reply by RedShyGuy** [reply]
Nope that is not needed with this patch.

**Reply by SirEnder125** [reply]
Ok. Thanks a lot! :D I'll fix the error I am currently getting and then try it out.
It seems that `make` thinks that the Makefile is in it's original position (in
downloads folder) and searches for my source code there... annoying.

**Reply by RedShyGuy** [reply]
What error is it showing you?

**Reply by SirEnder125** [reply]
This: `No rule to make target: 'C:\Users\######\Downloads\CTRPluginFramework-BlankTemplate-0.5.1\Sources\DumpRestore.cpp' needed by DumpRestore.o`

Why is it checking that directory when I am running make from C:\Users\\######\source\repos\3dsPlugin?

**Reply by RedShyGuy** [reply]
delete your Build folder and try again

**Reply by SirEnder125** [reply]
Ok

**Reply by SirEnder125** [reply]
Bruh! It worked. Thanks!! :D

**Reply by RedShyGuy** [reply]
Nice!

**Reply by SirEnder125** [reply]
Testing the cheat...
It works!!
Oops, I just placed a tree right in front of my house...
I guess the next cheat I need is pickup seeder/deleter. lol
